### PR TITLE
provider/aws: tag the spot instance.

### LINKED ISSUE #3263: aws_spot_instance_request tags won't apply to instance
I have an `aws_spot_instance_request` like this:

``` hcl
resource "aws_spot_instance_request" "seleniumgrid" {
    ami = "${var.amiPuppet}"
    key_name = "${var.key}"
    instance_type = "c3.4xlarge"
    subnet_id = "${var.subnet}"
    vpc_security_group_ids = [ "${var.securityGroup}" ]
    user_data = "${template_file.userdata.rendered}"
    wait_for_fulfillment = true
    spot_price = "${var.price}"
    availability_zone = "${var.zone}"
    instance_initiated_shutdown_behavior = "terminate"
    root_block_device {
      volume_size = 100
      volume_type = "gp2"
    }
    tags {
      Name = "${var.name}.${var.domain}"
      Provider = "Terraform"
      CA_TEAM = "${var.team}"
      CA_ROLE = "${var.role}"
      CA_SERVICE = "${var.service}"
    }
}
```

The tags are being applied only to the spot request itself, not to the underlying instance. Is this an expected behavior? How can I change this?

Thanks!

### PR DESCRIPTION
**Pull request by fanxu**
Fix #3263

### CODE REVIEW

**Comment by phinze** [comment]
Thanks for this! Can we add an acceptance test (or enhance an existing test) to verify that tags are set on the instance and cover this behavior?

**Comment by caarlos0** [comment]
Nice

**Comment by dmuysson** [comment]
We could really use this capability - hoping it gets added soon!

File: builtin/providers/aws/resource_aws_spot_instance_request.go [hunk 1]
```diff
@@ -140,11 +140,35 @@ func resourceAwsSpotInstanceRequestCreate(d *schema.ResourceData, meta interface
...
+		d.Set("spot_instance_id", sir.InstanceId)
+
+		// Wait for the spot instance running
+		log.Printf(
+			"[DEBUG] Waiting for instance (%s) to become running",
+			*sir.InstanceId)
+
+		stateConf := &resource.StateChangeConf{
+			Pending:    []string{"pending"},
+			Target:     "running",
```
Comment (crielly360) [inline builtin/providers/aws/resource_aws_spot_instance_request.go]: This line fails on compile for me, OSX and Linux

**Comment by harlow** [comment]
+1

**Comment by johnnyshields** [comment]
Not to split hairs here but there is technically a difference between tagging the SpotInstanceRequest itself versus tagging the EC2 Instance which it generates. Maybe using a different attribute like `instance_tags` for the instance (while keeping `tags` for the SIR) would be more accurate?

**Comment by johnnyshields** [comment]
Another comment: what happens with persistent requests when the instance is terminated and a new instance is later auto-created, presumably the new instance will no longer have the tags?

**Comment by fctucker** [comment]
+1 for this pull request to be accepted

**Comment by johnnyshields** [comment]
@fctucker one thing to note here is that with persistent spot requests, if the bid price is hit and the instance stops and starts again, the tags will be not be added to the newly spawned instance.

While this PR can behave as a stop-gap, I think everyone who wants support for this should ask AWS to support PropagateAtLaunch for SpotInstanceRequest tags in the same way it works for AutoScalingGroup tags, as that's a much better solution.

**Comment by jedi4ever** [comment]
👍 to get this merged  ; ideally it's added as an option = tag_instance true/false with default behaviour to false to be backwards compatible

**Comment by EntilZha** [comment]
Recently ran into this as well, would be great to get merged

**Comment by stack72** [comment]
FWIW, I think that @johnnyshields is correct here - I am going to open a feature request to AWS about this - feels like `propagate_at_launch` would be a good thing to have

**Comment by stack72** [comment]
just opened this https://github.com/aws/aws-sdk-go/issues/804

**Comment by stack72** [comment]
FYI, the tests for this PR work as expected:

```
% make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSSpotInstanceRequest_'                2 ↵
==> Checking that code complies with gofmt requirements...
/Users/stacko/Code/go/bin/stringer
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/08/15 13:22:41 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSSpotInstanceRequest_ -timeout 120m
=== RUN   TestAccAWSSpotInstanceRequest_basic
--- PASS: TestAccAWSSpotInstanceRequest_basic (147.58s)
=== RUN   TestAccAWSSpotInstanceRequest_withBlockDuration
--- PASS: TestAccAWSSpotInstanceRequest_withBlockDuration (134.29s)
=== RUN   TestAccAWSSpotInstanceRequest_vpc
--- PASS: TestAccAWSSpotInstanceRequest_vpc (174.33s)
=== RUN   TestAccAWSSpotInstanceRequest_SubnetAndSG
--- PASS: TestAccAWSSpotInstanceRequest_SubnetAndSG (167.04s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	623.262s
```

I made one modification, I check to see if the instance has the tags as expected

```
func testAccCheckAWSSpotInstanceRequest_InstanceTags(
	sir *ec2.SpotInstanceRequest) resource.TestCheckFunc {
	return func(s *terraform.State) error {
		conn := testAccProvider.Meta().(*AWSClient).ec2conn
		resp, err := conn.DescribeInstances(&ec2.DescribeInstancesInput{
			InstanceIds: []*string{sir.InstanceId},
		})
		if err != nil {
			if ec2err, ok := err.(awserr.Error); ok && ec2err.Code() == "InvalidInstanceID.NotFound" {
				return fmt.Errorf("Spot Instance not found")
			}
			return err
		}

		// If nothing was found, then return no state
		if len(resp.Reservations) == 0 {
			return fmt.Errorf("Spot Instance not found")
		}

		instance := resp.Reservations[0].Instances[0]

		var tagMatch bool
		for _, t := range instance.Tags {
			if *t.Value == "terraform-test" {
				tagMatch = true
			}
		}

		if !tagMatch {
			return fmt.Errorf("Error in matching Spot Instance Tags, expected 'terraform-test', got %s", instance.Tags[0].Value)
		}

		return nil
	}
}
```

**Comment by stack72** [comment]
We can wait and see what AWS has to say before merging this as a stop-gap

**Comment by radeksimko** [comment]
It would be super helpful to document that tags are only attached/detached if `wait_for_fulfillment = true` (default is `false`, so **by default** Terraform **won't** assign/change any tags). It may be obvious for us who know the implementation details behind it, but it may not be for a common user.

**Comment by elatt** [comment]
AWS now supports tagging natively for spot instances/fleets so this work-around should no longer be necessary if terraform updates its API params: https://aws.amazon.com/about-aws/whats-new/2017/07/tag-your-spot-fleet-ec2-instances/

**Comment by johnnyshields** [comment]
This PR should be closed and we should raise a new PR to support it natively.

**Comment by radeksimko** [comment]
Closing for the reasons mentioned. There's a new PR in the new repository also: https://github.com/terraform-providers/terraform-provider-aws/pull/1366

**Comment by c0d5x** [comment]
PR #1366 does *NOT* fix this, instead if fixes the spot *fleet* requests. This PR is created for spot *instances*, so it is still needed.

**Pull request closed without merging**